src/main/java/com/icthh/xm/tmf/ms/activation/utils/TenantUtils.java

```diff
@@ -38,6 +38,17 @@ public void doInTenantContext(Task task, String tenant) {
         }
     }
 
+    @SneakyThrows
+    public <R> R doInTenantContext(TaskWithResult<R> task, String tenant) {
+        try {
+            init(tenant);
+            return task.doWork();
+        } finally {
+            destroy();
+        }
+    }
+
+
     private void init(String tenantKey) {
         TenantContextUtils.setTenant(tenantContextHolder, tenantKey);
```

Comment:
Do not use Set tenant to switch tenant context. 
Use defined mechanism for logic execution for defined tenant context: 
```
tenantContextHolder.getPrivilegedContext().execute( ... 
```